Code/Framework/AzCore/AzCore/DOM/DomPatch.h

```diff
@@ -0,0 +1,191 @@
+        explicit PatchOperation(Path path, Type type);
+
+        static PatchOperation AddOperation(Path destinationPath, Value value);
+        static PatchOperation RemoveOperation(Path pathToRemove);
+        static PatchOperation ReplaceOperation(Path destinationPath, Value value);
+        static PatchOperation CopyOperation(Path destinationPath, Path sourcePath);
+        static PatchOperation MoveOperation(Path destinationPath, Path sourcePath);
+        static PatchOperation TestOperation(Path testPath, Value value);
+
+        PatchOperation& operator=(const PatchOperation&) = default;
+        PatchOperation& operator=(PatchOperation&&) = default;
+
+        bool operator==(const PatchOperation& rhs) const;
+        bool operator!=(const PatchOperation& rhs) const;
+
+        Type GetType() const;
+        void SetType(Type type);
+
+        const Path& GetDestinationPath() const;
+        void SetDestinationPath(Path path);
+
+        const Value& GetValue() const;
+        void SetValue(Value value);
+
+        const Path& GetSourcePath() const;
+        void SetSourcePath(Path path);
+
+        AZ::Outcome<Value, AZStd::string> Apply(Value rootElement) const;
+        PatchOutcome ApplyInPlace(Value& rootElement) const;
+
+        Value GetDomRepresentation() const;
+        static AZ::Outcome<PatchOperation, AZStd::string> CreateFromDomRepresentation(Value domValue);
+
+        AZ::Outcome<PatchOperation, AZStd::string> GetInverse(Value stateBeforeApplication) const;
+
+    private:
+        struct PathContext
+        {
+            Value& m_value;
+            PathEntry m_key;
+        };
+
+        static constexpr AZ::u8 DefaultExistenceCheck = 0x0;
+        static constexpr AZ::u8 VerifyFullPath = 0x1;
+        static constexpr AZ::u8 AllowEndOfArray = 0x2;
+
+        static AZ::Outcome<PathContext, AZStd::string> LookupPath(
+            Value& rootElement, const Path& path, AZ::u8 existenceCheckFlags = DefaultExistenceCheck);
+
+        PatchOutcome ApplyAdd(Value& rootElement) const;
+        PatchOutcome ApplyRemove(Value& rootElement) const;
+        PatchOutcome ApplyReplace(Value& rootElement) const;
+        PatchOutcome ApplyCopy(Value& rootElement) const;
+        PatchOutcome ApplyMove(Value& rootElement) const;
+        PatchOutcome ApplyTest(Value& rootElement) const;
+
+        Path m_domPath;
+        Type m_type;
+        AZStd::variant<AZStd::monostate, Value, Path> m_value;
+    };
+
+    class Patch;
+
+    //! The current state of a Patch application operation.
+    struct PatchApplicationState
+    {
+        //! The patch being applied.
+        const Patch* m_patch = nullptr;
+        //! The last operation attempted.
+        const PatchOperation* m_lastOperation = nullptr;
+        //! The outcome of the last operation, may be overridden to produce a different failure outcome.
+        PatchOperation::PatchOutcome m_outcome;
```

Comment:
Nitpick: for classes that can have a large number of instance it's recommended to by default sort member values from largest to smallest to avoid needles padding.

Reply:
Good call, reordered.

Reply:
You may have forgotten to do the actual reordering.

Reply:
Hah I reordered another struct... should be in proper `Outcome<AZStd::string>` -> pointers -> bool order now.